src/PostgREST/App.hs

```diff
@@ -9,430 +9,664 @@ Some of its functionality includes:
-
-        readSqlParts s t =
-          let
-            readReq = readRequest s t maxRows (dbRelations dbStructure) apiRequest
-            returnings :: ReadRequest -> Either Response [FieldName]
-            returnings rr = Right (returningCols rr [])
-          in
-          (,,,) <$>
-          (readRequestToQuery <$> readReq) <*>
-          (readRequestToCountQuery <$> readReq) <*>
-          (binaryField contentType rawContentTypes returnsScalar =<< readReq) <*>
-          (returnings =<< readReq)
-
-        mutateSqlParts s t =
-          let
-            readReq = readRequest s t maxRows (dbRelations dbStructure) apiRequest
-            mutReq = mutateRequest s t apiRequest (tablePKCols dbStructure s t) =<< readReq
-          in
-          (,) <$>
-          (readRequestToQuery <$> readReq) <*>
-          (mutateRequestToQuery <$> mutReq)
-
-responseContentTypeOrError :: [ContentType] -> [ContentType] -> Action -> Target -> Either Response ContentType
-responseContentTypeOrError accepts rawContentTypes action target = serves contentTypesForRequest accepts
-  where
-    contentTypesForRequest = case action of
-      ActionRead _       ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-                             ++ rawContentTypes
-      ActionCreate       ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-      ActionUpdate       ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-      ActionDelete       ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-      ActionInvoke _     ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-                             ++ rawContentTypes
-                             ++ [CTOpenAPI | tpIsRootSpec target]
-      ActionInspect _    ->  [CTOpenAPI, CTApplicationJSON]
-      ActionInfo         ->  [CTTextCSV]
-      ActionSingleUpsert ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-    serves sProduces cAccepts =
-      case mutuallyAgreeable sProduces cAccepts of
-        Nothing -> Left . errorResponseFor . ContentTypeError . map toMime $ cAccepts
-        Just ct -> Right ct
-
-{-
-  | If raw(binary) output is requested, check that ContentType is one of the admitted rawContentTypes and that
-  | `?select=...` contains only one field other than `*`
--}
-binaryField :: ContentType -> [ContentType] -> Bool -> ReadRequest -> Either Response (Maybe FieldName)
-binaryField ct rawContentTypes isScalarProc readReq
-  | isScalarProc =
-      if ct `elem` rawContentTypes
-        then Right $ Just "pgrst_scalar"
-        else Right Nothing
-  | ct `elem` rawContentTypes =
-      let fieldName = headMay fldNames in
-      if length fldNames == 1 && fieldName /= Just "*"
-        then Right fieldName
-        else Left . errorResponseFor $ BinaryFieldError ct
-  | otherwise = Right Nothing
+postgrestResponse
+  :: AppConfig
+  -> Maybe DbStructure
+  -> SQL.Pool
+  -> UTCTime
+  -> Wai.Request
+  -> Handler IO Wai.Response
+postgrestResponse conf maybeDbStructure pool time req =
```

Comment:
I feel like this whole function should really be part of the `postgrest` function. Once the middleware is moved out of there, it should not increase nesting that much.

Both of those functions taken together just take all the different pieces that are done before and after handling a request and stitch them together. Right now they are spread across two functions, I think they fit better together.

Reply:
There is some point in separating the IORef stuff from the actual, `pure` response function. At this point, merging the functions without changing the behavior of https://github.com/monacoremo/postgrest/blob/a6ba8784e9096eee2e30f3c049ec7eeec9886a7c/src/PostgREST/App.hs#L90 will be tricky, so I'd leave it unchanged in this PR; let's review this in a separate one.

We'd also need an extra level of indentation, as we need to run everything in `runExceptT` to avoid the nested case analysis.